File: 155-min-stack/min-stack.js

class MinStack {
    constructor() {
        this.stack = [];     // Main stack
        this.minStack = [];  // Stack to track minimum values
    }

    push(val) {
        this.stack.push(val);
        if (this.minStack.length === 0 || val <= this.minStack[this.minStack.length - 1]) {
            this.minStack.push(val);
        }
    }

    pop() {
        if (this.stack.length > 0) {
            const popped = this.stack.pop();
            if (popped === this.minStack[this.minStack.length - 1]) {
                this.minStack.pop();
            }
        }
    }

    top() {
        if (this.stack.length > 0) {
            return this.stack[this.stack.length - 1];
        }
        return null;
    }

    getMin() {
        if (this.minStack.length > 0) {
            return this.minStack[this.minStack.length - 1];
        }
        return null;
    }
}


/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 155-min-stack/min-stack.py
class MinStack:
    

    def __init__(self):
        self.stack = []
        self.minstack = []

    def push(self, val: int) -> None:
    
        self.stack.append(val)
        #append the curr val to minstack if - 
        # - min stack is empty 
        # - min stack top <= curr val
        if not self.minstack or val <= self.minstack[-1]:
            self.minstack.append(val)
        

    def pop(self) -> None:
        if self.stack:
            popped = self.stack.pop()

            if self.minstack and popped == self.minstack[-1]:
                self.minstack.pop()


    def top(self) -> int:
        if self.stack:
            return self.stack[-1]

    def getMin(self) -> int:
        return self.minstack[-1] if self.minstack else None


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()